### PR TITLE
Add support for PyPy and Python 3.4

### PR DESCRIPTION
**Pull request by jamadden**
This PR is the one referenced by and closing #3 by adding support for PyPy (and Python 3.4 as a bonus). 

All tests continue to pass under 2.6 and 2.7. 

The pure-python code is basically a straightforward port of the C version. Of the 47 tests, 46 pass;  ` Acquisition.tests.test___parent__aq_parent_circles` currently raises `AttributeError` instead of `RuntimeError` in its recursive-parent cases, though I'm not sure why that is, nor am I sure if it's actually a problem or just a difference. I'll keep looking, but any insight is appreciated:

```python
Failed example:
    Acquisition.aq_acquire(y, 'non_existant_attr')
Expected:
    Traceback (most recent call last):
    ...
    RuntimeError: Recursion detected in acquisition wrapper
Got:
    Traceback (most recent call last):
      File "/opt/local/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/doctest.py", line 1318, in __run
        compileflags, 1), test.globs)
      File "<doctest Acquisition.tests.test___parent__aq_parent_circles[12]>", line 1, in <module>
        Acquisition.aq_acquire(y, 'non_existant_attr')
      File "/Acquisition/src/Acquisition/__init__.py", line 485, in aq_acquire
        containment=containment)
      File "/Acquisition/src/Acquisition/__init__.py", line 364, in aq_acquire
        containment=containment)
      File "/Acquisition/src/Acquisition/__init__.py", line 277, in _Wrapper_findattr
        containment=containment)
      File "/Acquisition/src/Acquisition/__init__.py", line 160, in _Wrapper_acquire
        explicit=explicit, containment=containment)
      File "/Acquisition/src/Acquisition/__init__.py", line 279, in _Wrapper_findattr
        raise AttributeError(orig_name)
    AttributeError: non_existant_attr
```

I suspect that some of the special functions aren't correctly proxied, but I'm not sure because I don't think they all have test cases. Perhaps another PR could add test cases and fix any differences? (*EDIT*: I'm about done writing test cases for proxying so I'll probably add them to this PR.)

This code contains a workaround for zopefoundation/ExtensionClass#3, so a coordinated release is probably not necessary, albeit nice.

### CODE REVIEW

**Comment by tseaver** [comment]
Can you please diagnose / fix the [Travis failures](https://travis-ci.org/zopefoundation/Acquisition/builds/55094255)?  Just make a new commit with the fix and re-push your branch -- Travis will re-do the test run for this PR at that point.

**Comment by jamadden** [comment]
Yay, the build is green! The blocker turned out to be the issue in zopefoundation/ExtensionClass#3 (which needs some tox love that I'm not sure how to give).

**Comment by jamadden** [comment]
@tseaver and/or @davisagli, I would welcome any feedback. I apologize for opening the PR a little early, but I was having trouble tracking down that last failure, and I was hoping that the travis output might help me find it, or failing that, more eyeballs might make the bug more shallow. Also, I didn't realize the state that the proxy tests were in until later; nothing except that one test was failing so I didn't know there was more to do.

It's a big diff, I know, but hopefully it's fairly straightforward. As far as tests go, I didn't modify the existing tests anymore than necessary to get them to run under Python3. I also added a bunch for the proxy/wrapping behaviour that wasn't covered before (which does expose some interesting quirks in the C implementation that I've faithfully copied).

As far as the implementation goes, I know it's not exactly idiomatic Python, but it's mostly a simple port of the existing C code. I'm hoping this helps ease maintenance of the parallel implementations, as well as catching any corner cases still not covered by tests. 

One of the most interesting or controversial quirks is the behaviour of the C proxy wrappers when it comes to comparisons. Specifically, it only delegates to `__cmp__`, which is not really applicable in Python 3. On the principle of least surprise for upgraders, I kept that behaviour intact across all platforms (I can imagine breakage if `__eq__` and friends started getting called suddenly sometimes, if they're ever even implemented). Maybe there's something better?

I currently only know the Python C API well enough to read it, but if you want me to take a shot at porting the C extension to Python 3 I can give that a try.

File: setup.py [hunk 1]
```diff
@@ -22,6 +24,21 @@
...
+    ext_modules=[Extension("Acquisition._Acquisition",
+                           [os.path.join('src', 'Acquisition',
+                                         '_Acquisition.c')],
+                           include_dirs=['include', 'src']),
+                 ]
+
+
 setup(
     name='Acquisition',
     version='4.1',
```
Comment (davisagli) [inline setup.py]: This should change to 4.2.dev.0 to match the minor version bump in the changelog.

File: src/Acquisition/__init__.py [hunk 2]
```diff
@@ -1,8 +1,833 @@
...
+def _has__of__(obj):
+    """Check whether an object has an __of__ method for returning itself
+    in the context of a container."""
+    # Note specifically this is a type check, not duck typing, or
+    # we get into cycles
+    return isinstance(obj, ExtensionClass.Base)
+
+
+def _apply_filter(predicate, inst, name, result, extra, orig):
+    return predicate(orig, inst, name, result, extra)
```
Comment (davisagli) [inline src/Acquisition/__init__.py]: Why not just inline this function?

File: src/Acquisition/tests.py [hunk 3]
```diff
@@ -1620,25 +1655,31 @@ def test_cant_persist_acquisition_wrappers_classic():
     >>> pickler = cPickle.Pickler(file, 1)
 
     >>> def persistent_id(obj):
+    ...     if not hasattr(obj, '_p_oid'): return None
```
Comment (davisagli) [inline src/Acquisition/tests.py]: Is this due to a difference between the pure-Python and C implementations of `Persistent`?

Comment (davisagli) [inline src/Acquisition/tests.py]: Er, of course not; no Persistent in sight. So I guess this is to deal with some difference in how pypy or Python 3 handles an AttributeError in persistent_id?

**Comment by davisagli** [comment]
I skimmed through the changes and at first glance it looks fine. I haven't done a detailed comparison with the C implementation but it's a good sign that the tests are passing. I agree it makes sense to keep the Python implementation close to a literal translation of the C, at least as long as we want to maintain the C implementation too.

File: src/Acquisition/__init__.py (hunk 2 above)
Comment (jamadden) [inline src/Acquisition/__init__.py]: Basically, because that's the way the C implementation does it. The order of arguments to the user-provided predicate function is *different* from the order of arguments passed to the helper as written in the C (and Python code) so calling the predicate directly would introduce a (gratuitous?) difference that's likely to get mistakenly changed in maintenance.  Plus, when the helper function is called, the caller's variable names are not really meaningful in terms of the arguments to the predicate function; the helper is a layer of abstraction that, to me at least,  helps smooth over the impedance mismatch.

File: src/Acquisition/tests.py (hunk 3 above)
Comment (jamadden) [inline src/Acquisition/tests.py]: Exactly. The `cPickle` implementation used in CPython 2.x  and 3.x's C-accelerated implementation simply invokes `PyObject_Call` on the user-provided function and checks its return value; it *does not* check for exceptions being raised (as far as I can tell; I'm not a C API expert), so they are silently ignored. In contrast, PyPy (or the non-C-accelerated `pickle` module) make the call at the Python level, so exceptions that the `persistent_id` function raise are propagated all the way up, breaking this test. Hence the need to keep `persistent_id` from raising when called with a "non-persistent" object.

File: src/Acquisition/__init__.py [hunk 4]
```diff
@@ -1,8 +1,833 @@
...
+# Helper functions
+###
+
+
+def _has__of__(obj):
+    """Check whether an object has an __of__ method for returning itself
+    in the context of a container."""
+    # Note specifically this is a type check, not duck typing, or
+    # we get into cycles
+    return isinstance(obj, ExtensionClass.Base)
```
Comment (davisagli) [inline src/Acquisition/__init__.py]: The C implementation of has__of__ (actually a macro in ExtensionClass.h) checks not only that obj is an ExtensionClass instance, but also that its tp_descr_get slot filled (which for ExtensionClasses is equivalent to it having __of__ defined).

I'm running some Zope tests in PyPy using this branch of Acquisition, and hitting errors when I try to acquire something from an acquisition wrapper around a class that is an ExtensionClass but does not define __of__:

```
Traceback (most recent call last):
  File "/Users/davisagli/src/pypy-2.5.1-osx64/lib-python/2.7/unittest/case.py", line 320, in run
    self.setUp()
  File "/Users/davisagli/Code/Zope/develop/AccessControl/src/AccessControl/tests/testOwned.py", line 207, in setUp
    self.uf._doAddUser('user1', 'xxx', ['role1'], [])
  File "/Users/davisagli/Code/Zope/develop/AccessControl/src/AccessControl/userfolder.py", line 410, in _doAddUser
    self.data[name] = User(name, password, roles, domains)
  File "/Users/davisagli/Code/Zope/develop/Acquisition/src/Acquisition/__init__.py", line 327, in __getattribute__
    True, type(self)._IS_IMPLICIT, False, False)
  File "/Users/davisagli/Code/Zope/develop/Acquisition/src/Acquisition/__init__.py", line 262, in _Wrapper_findattr
    result = result.__of__(wrapper)
  File "/Users/davisagli/Code/Zope/eggs/persistent-4.0.8-py2.7.egg/persistent/persistence.py", line 238, in __getattribute__
    return _OGA(self, name)
AttributeError: 'PersistentMapping' object has no attribute '__of__'
```

Comment (davisagli) [inline src/Acquisition/__init__.py]: This seems to work:

```
def _has__of__(obj):
    """Check whether an object has an __of__ method for returning itself
    in the context of a container."""
    return isinstance(obj, ExtensionClass.Base) and hasattr(type(obj), '__of__')
```

Comment (jamadden) [inline src/Acquisition/__init__.py]: Good point about the C implementation. In pure Python wasn't clear to me whether it would be likely to get in the state where `isinstance(o,Base) == True` yet `getattr(o,'__of__')` would raise---that takes some special circumstances (does Zope somewhere mixin `Base` as a base-class of PersistentMapping post definition?)

I'll try to set up a testcase and make that change.

Comment (davisagli) [inline src/Acquisition/__init__.py]: It's not changing `__bases__`. AccessControl uses the PersistentMapping class from Persistence (sic) which mixes in ExtensionClass.Base via this: https://github.com/zopefoundation/Persistence/blob/master/src/Persistence/__init__.py#L7

David Glick
(mobile)

Comment (jamadden) [inline src/Acquisition/__init__.py]: Ok, I see it now. Thank you.

File: src/Acquisition/tests.py [hunk 5]
```diff
@@ -1663,7 +1704,7 @@ def test_cant_persist_acquisition_wrappers_newstyle():
 
     Check custom pickler:
 
-    >>> from cStringIO import StringIO
+    >>> from io import BytesIO as StringIO
```
Comment (tseaver) [inline src/Acquisition/tests.py]: Better to rename the usage points `BytesIO` (for clarity), than to alias it here.

Comment (jamadden) [inline src/Acquisition/tests.py]: Will do. I was just trying to keep the diff minimal.

**Comment by tseaver** [comment]
This is quite an impressive piece of work -- kudos!

I'm a little worried about test coverage, especially for the new Python implementation (I suppose that means for the existing C implementation, too).

**Comment by jamadden** [comment]
I was a bit worried about the test coverage too, but the numbers actually look pretty decent:


```
$ python --version
Python 2.7.8 (10f1b29a2bd2, Feb 02 2015, 21:22:55)
[PyPy 2.5.0 with GCC 4.2.1 Compatible Apple LLVM 6.0 (clang-600.0.56)]
$ coverage run --source=Acquisition bin/test
Running zope.testrunner.layer.UnitTests tests:
  Set up zope.testrunner.layer.UnitTests in 0.000 seconds.
  Ran 61 tests with 0 failures, 0 errors and 0 skipped in 1.512 seconds.
Tearing down left over layers:
  Tear down zope.testrunner.layer.UnitTests in 0.000 seconds.
$ coverage report
Name                         Stmts   Miss  Cover
------------------------------------------------
src/Acquisition/__init__       412     37    91%
src/Acquisition/interfaces      14      0   100%
src/Acquisition/tests          416     25    94%
------------------------------------------------
TOTAL                          842     62    93%
```

I can look at the statement coverage in greater depth and try to improve the numbers.

**Comment by jamadden** [comment]
I brought the coverage up to 98% for `__init__`, which is most of the low-hanging fruit, I think, without adding too many `no cover` lines, I hope.

**Comment by jamadden** [comment]
I got the Python coverage up to 99%, leaving only one branch (two lines, 120/121) uncovered without a `no cover` remark. It may be possible to get there without exceptional circumstances, but I've given up thinking about it for today :) One branch I did add a `no cover` to I'm pretty sure really is impossible to get to and is a result of the C code being defensive. The upshot is that the added tests are (mostly) applicable to both the C and Python implementations, so coverage overall is way up.

**Comment by jamadden** [comment]
I had to resort to a Python-implementation-specific testcase to get that last branch covered, but I'm pleased to report 100% coverage now.

```
$ coverage report
Name                         Stmts   Miss  Cover
------------------------------------------------
src/Acquisition/__init__       403      0   100%
src/Acquisition/interfaces      14      0   100%
src/Acquisition/tests          434     27    94%
------------------------------------------------
TOTAL                          851     27    97%
```

**Comment by tseaver** [comment]
Thank you for the great work!
